src/pipx/commands/install.py

```diff
@@ -117,3 +119,88 @@ def install(
 
     # Any failure to install will raise PipxError, otherwise success
     return EXIT_CODE_OK
+
+
+def extract_venv_metadata(spec_metadata_file: Path) -> Iterator[PipxMetadata]:
+    """Extracts venv metadata from spec metadata file."""
+    with open(spec_metadata_file) as spec_metadata_fh:
+        try:
+            spec_metadata_dict = json.load(spec_metadata_fh, object_hook=_json_decoder_object_hook)
+        except json.decoder.JSONDecodeError as exc:
+            raise PipxError("The spec metadata file is an invalid JSON file.") from exc
+
+        if not ("venvs" in spec_metadata_dict and len(spec_metadata_dict["venvs"])):
+            raise PipxError("No packages found in the spec metadata file.")
+
+        venvs_metadata_dict = spec_metadata_dict["venvs"]
+
+        if not isinstance(venvs_metadata_dict, dict):
+            raise PipxError("The spec metadata file is invalid.")
+
+        for package_path_name in venvs_metadata_dict:
+            venv_dir = paths.ctx.venvs.joinpath(package_path_name)
+            venv_metadata = PipxMetadata(venv_dir, read=False)
+            venv_metadata.from_dict(venvs_metadata_dict[package_path_name]["metadata"])
+            yield venv_metadata
+
+
+def generate_package_spec(package_info: PackageInfo) -> str:
+    """Generate more precise package spec from package info."""
+    if not package_info.package_or_url:
+        raise PipxError(f"The package_or_url of {package_info.package} is empty.")
+
+    if package_info.package == package_info.package_or_url:
+        return f"{package_info.package}=={package_info.package_version}"
+    return package_info.package_or_url
+
+
+def install_all(
+    spec_metadata_file: Path,
+    local_bin_dir: Path,
+    local_man_dir: Path,
+    pip_args: List[str],
+    venv_args: List[str],
+    verbose: bool,
+    *,
+    force: bool,
+) -> ExitCode:
+    """Returns pipx exit code."""
+    venv_container = VenvContainer(paths.ctx.venvs)
+
+    for venv_metadata in extract_venv_metadata(spec_metadata_file):
+        # Install the main package
+        main_package = venv_metadata.main_package
+        venv_dir = venv_container.get_venv_dir(f"{main_package.package}{main_package.suffix}")
+        install(
+            venv_dir,
+            None,
+            [generate_package_spec(main_package)],
+            local_bin_dir,
+            local_man_dir,
+            venv_metadata.python_version,
+            pip_args,
+            venv_args,
+            verbose,
+            force=force,
+            reinstall=False,
```

Comment:
What if we have a package already installed?

Reply:
If the user does not specify the `--force` flag, everything will be kept as it is. And when the user specify the `--force` flag, the package will be overwrite.

I specify the `reinstall` as `False` here, because it means to first uninstall the package and then install it again, that is not an ordinary scenario in `install-all` I think.